x/ccv/provider/ibc_middleware.go

```diff
@@ -0,0 +1,240 @@
+// and if the received IBC coin is whitelisted. In such instances,
+// it appends the coin to the consumer's chain allocation record
+func (im IBCMiddleware) OnRecvPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	relayer sdk.AccAddress,
+) exported.Acknowledgement {
+	// executes the IBC transfer OnRecv logic
+	ack := im.app.OnRecvPacket(ctx, packet, relayer)
+
+	// execute the middleware logic only if the sender is a consumer chain
+	consumerID, err := im.keeper.IdentifyConsumerChainIDFromIBCPacket(ctx, packet)
+	if err != nil {
+		return ack
+	}
+
+	// Note that inside the above if condition statement,
+	// we know that the IBC transfer succeeded. That entails
+	// that the packet data is valid and can be safely
+	// deserialized without checking errors.
+	if ack.Success() {
+		// extract the coin info received from the packet data
+		var data ibctransfertypes.FungibleTokenPacketData
+		_ = types.ModuleCdc.UnmarshalJSON(packet.GetData(), &data)
+
+		// check if the recipient is the consumer reward's pool address
+		receiver, _ := sdk.AccAddressFromBech32(data.Receiver)
+		if receiver.String() != im.keeper.GetConsumerRewardsPoolAddressStr(ctx) {
+			return ack
+		}
+
+		coinAmt, _ := math.NewIntFromString(data.Amount)
+		coinDenom := GetProviderDenom(data.Denom, packet)
+
+		// verify that the coin's denom is a whitelisted consumer denom,
+		// and if so, adds it to the consumer chain rewards allocation,
+		// otherwise the prohibited coin just stays in the pool forever.
+		if im.keeper.ConsumerRewardDenomExists(ctx, coinDenom) {
+			alloc := im.keeper.GetConsumerRewardsAllocation(ctx, consumerID)
+			alloc.Rewards = alloc.Rewards.Add(
+				sdk.NewDecCoinsFromCoins(sdk.Coin{
+					Denom:  coinDenom,
+					Amount: coinAmt,
+				})...)
+			im.keeper.SetConsumerRewardsAllocation(ctx, consumerID, alloc)
+		}
+	}
+
+	return ack
+}
+
+// OnAcknowledgementPacket implements the IBCMiddleware interface
+// If fees are not enabled, this callback will default to the ibc-core packet callback
+func (im IBCMiddleware) OnAcknowledgementPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	acknowledgement []byte,
+	relayer sdk.AccAddress,
+) error {
+	// call underlying app's OnAcknowledgementPacket callback.
+	return im.app.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer)
+}
+
+// OnTimeoutPacket implements the IBCMiddleware interface
+// If fees are not enabled, this callback will default to the ibc-core packet callback
+func (im IBCMiddleware) OnTimeoutPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	relayer sdk.AccAddress,
+) error {
+	// call underlying app's OnTimeoutPacket callback.
+	return im.app.OnTimeoutPacket(ctx, packet, relayer)
+}
+
+// SendPacket implements the ICS4 Wrapper interface
+func (im IBCMiddleware) SendPacket(
+	sdk.Context,
+	*capabilitytypes.Capability,
+	string,
+	string,
+	clienttypes.Height,
+	uint64,
+	[]byte,
+) (uint64, error) {
+	return 0, nil
```

Comment:
Panic instead. This middleware doesn't have an ICS4 Wrapper. This means that none of the `SendPacket`, `WriteAcknowledgement`, and `GetAppVersion` methods are expected to go through the middleware. If that happens, something is clearly wrong (probably with the wiring) and we should panic.